File: perf/benchmarks/micro/quicksort/quicksort.go
// Hand-written quicksort — micro-benchmark (Go).
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func quicksort(a []int64, lo, hi int64) {
	for lo < hi {
		mid := (lo + hi) >> 1
		pivot := a[mid]
		a[mid], a[hi] = a[hi], a[mid]
		i := lo - 1
		for j := lo; j < hi; j++ {
			if a[j] <= pivot {
				i++
				a[i], a[j] = a[j], a[i]
			}
		}
		i++
		a[i], a[hi] = a[hi], a[i]
		if (i - lo) < (hi - i) {
			quicksort(a, lo, i-1)
			lo = i + 1
		} else {
			quicksort(a, i+1, hi)
			hi = i - 1
		}
	}
}

func main() {
	n := int64(1_500_000)
	if len(os.Args) > 1 {
		v, _ := strconv.ParseInt(os.Args[1], 10, 64)
		n = v
	}
	a := make([]int64, n)
	seed := int64(1234567)
	const mask int64 = 0x7FFFFFFF
	for i := int64(0); i < n; i++ {
		seed = (seed*1103515245 + 12345) & mask
		a[i] = seed
	}
	t0 := time.Now()
	quicksort(a, 0, n-1)
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	var cs int64
	for k := int64(0); k < n; k += 1024 {
		cs ^= a[k]
	}
	fmt.Printf("CHECKSUM:%d\n", cs)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/micro/quicksort/quicksort.rs
// Hand-written quicksort — micro-benchmark (Rust).
use std::env;
use std::time::Instant;

fn quicksort(a: &mut [i64], mut lo: i64, mut hi: i64) {
    while lo < hi {
        let mid = (lo + hi) >> 1;
        let pivot = a[mid as usize];
        a.swap(mid as usize, hi as usize);
        let mut i = lo - 1;
        let mut j = lo;
        while j < hi {
            if a[j as usize] <= pivot {
                i += 1;
                a.swap(i as usize, j as usize);
            }
            j += 1;
        }
        i += 1;
        a.swap(i as usize, hi as usize);
        if (i - lo) < (hi - i) {
            quicksort(a, lo, i - 1);
            lo = i + 1;
        } else {
            quicksort(a, i + 1, hi);
            hi = i - 1;
        }
    }
}

fn main() {
    let args: Vec<String> = env::args().collect();
    let n: i64 = if args.len() > 1 { args[1].parse().unwrap_or(1_500_000) } else { 1_500_000 };
    let mut a: Vec<i64> = vec![0; n as usize];
    let mut seed: i64 = 1234567;
    let mask: i64 = 0x7FFFFFFF;
    for i in 0..n as usize {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & mask;
        a[i] = seed;
    }
    let t0 = Instant::now();
    quicksort(&mut a, 0, n - 1);
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    let mut cs: i64 = 0;
    let mut k: i64 = 0;
    while k < n {
        cs ^= a[k as usize];
        k += 1024;
    }
    println!("CHECKSUM:{}", cs);
    println!("ELAPSED_MS:{:.6}", ms);
}
